delfin/drivers/hitachi/vsp/vsp_stor.py

```diff
@@ -0,0 +1,295 @@
+        status = constants.StorageStatus.OFFLINE
+        if firmware_version is not None:
+            status = constants.StorageStatus.NORMAL
+        system_name = '%s_%s' % (self.rest_handler.device_model,
+                                 self.rest_handler.rest_host)
+
+        s = {
+            'name': system_name,
+            'vendor': 'Hitachi',
+            'description': 'Hitachi VSP Storage',
+            'model': str(self.rest_handler.device_model),
+            'status': status,
+            'serial_number': str(self.rest_handler.serial_number),
+            'firmware_version': str(firmware_version),
+            'location': '',
+            'raw_capacity': int(total_total),
+            'total_capacity': int(total_total),
+            'used_capacity': int(used_total),
+            'free_capacity': int(free_total)
+        }
+        return s
+
+    def list_storage_pools(self, context):
+        try:
+            pools_info = self.rest_handler.get_all_pools()
+            pool_list = []
+            pools = pools_info.get('data')
+            for pool in pools:
+                status = self.POOL_STATUS_MAP.get(
+                    pool.get('poolStatus'),
+                    constants.StoragePoolStatus.ABNORMAL
+                )
+                storage_type = constants.StorageType.BLOCK
+                total_cap = \
+                    int(pool.get('totalPoolCapacity')) * units.Mi
+                free_cap = int(
+                    pool.get('availableVolumeCapacity')) * units.Mi
+                used_cap = total_cap - free_cap
+                p = {
+                    'name': pool.get('poolName'),
+                    'storage_id': self.storage_id,
+                    'native_storage_pool_id': str(pool.get('poolId')),
+                    'description': 'Hitachi VSP Pool',
+                    'status': status,
+                    'storage_type': storage_type,
+                    'subscribed_capacity': int(total_cap),
+                    'total_capacity': int(total_cap),
+                    'used_capacity': int(used_cap),
+                    'free_capacity': int(free_cap),
+                }
+                pool_list.append(p)
+
+            return pool_list
+        except exception.DelfinException as err:
+            err_msg = "Failed to get pool metrics from hitachi vsp: %s" % \
+                      (six.text_type(err))
+            LOG.error(err_msg)
+            raise err
+        except Exception as e:
+            err_msg = "Failed to get pool metrics from hitachi vsp: %s" % \
+                      (six.text_type(e))
+            LOG.error(err_msg)
+            raise exception.InvalidResults(err_msg)
+
+    def list_volumes(self, context):
+        try:
+            volumes_info = self.rest_handler.get_all_volumes()
+
+            volume_list = []
+            volumes = volumes_info.get('data')
+            for volume in volumes:
+                if volume.get('emulationType') == 'NOT DEFINED':
```

Comment:
if a volume doesn't have emulation type, where will be the capacity or storage be accounted?

Reply:
in hitachi rest api of the volume,there always have emulationType

Reply:
Then what is this 'NOT DEFINED?